net: Remove unpacked structure references from packed structs

rlubos opened the pull request:
This PR proposes a solution for the "unaligned access to the packed struct member warning" issue within networking code.

The generic approach for the problem is similar in most cases covered in this PR:
* Replace `struct in_addr`/`struct in6_addr` within packed structs with raw buffers,
* Provide `_raw` variants of the most commonly used functions (IPv4/6 copy and compare) and use them where applicable,
* ~~For other cases, create a local copy of `struct in_addr`/`struct in6_addr` on stack, copy the IP address from the actual packet, and work with the local copy istead of referencing possibly unaligned address from the packet.~~

The only exception for the above scheme is the ieee802154 case - here IMO there's no need to use __packed qualifier for the structure that generated warnings, for the reason explained in the commit message.

The only warning left when building networking code is the following in the `mmu.c` file (this will likely cause CI failures):
```
/home/robert/repos/zephyrproject/zephyr/kernel/mmu.c: In function 'free_page_frame_list_put':
/home/robert/repos/zephyrproject/zephyr/kernel/mmu.c:333:42: warning: taking address of packed member of 'struct z_page_frame' may result in an unaligned pointer value [-Waddress-of-packed-member]
  333 |  sys_slist_append(&free_page_frame_list, &pf->node);
      | 
```

Fixes #16587

rlubos committed:
net: arp: Remove in_addr/in6_addr from packed net_arp_hdr struct

Replace unpacked in_addr/in6_addr structures with raw buffers in
net_arp_hdr struct, to prevent compiler warnings about unaligned
access.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: ip: Verify that in_addr/in6_addr struct sizes are correct

Make sure that in_addr/in6_addr structure size match the respective
binary IP address size with BUILD_ASSERT.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: ipv4: Remove in_addr from packed net_ipv4_hdr struct

Replace unpacked in_addr structures with raw buffers in net_ipv4_hdr
struct, to prevent compiler warnings about unaligned access.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: ipv6: Remove in6_addr from packed net_ipv6_hdr struct

Replace unpacked in6_addr structures with raw buffers in net_ipv6_hdr
struct, to prevent compiler warnings about unaligned access.

Remove __packed parameter from `struct net_6lo_context` since the
structure isn't really serialized.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: l2: ieee802154: Fix packed structure warning

`struct ieee802154_mpdu` access generates warnings related to unaligned
access to packed structure member. The structure itself however does not
need to be packed, since it is not mapped directly into the frame
buffer, but rather contains pointers to the structures representing
corresponding header regions in the frame. Those structures are
correctly defined as packed.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: icmpv6: Remove in6_addr from packed structs

Replace unpacked in6_addr structures with raw buffers in packed icmpv6
structs, to prevent compiler warnings about unaligned access.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

jukkar commented on subsys/net/l2/ethernet/arp.c:
```diff
@@ -535,11 +535,15 @@ static inline struct net_pkt *arp_prepare_reply(struct net_if *iface,
 {
+	struct in_addr src_ip;
+
+	memcpy(&src_ip, arp_hdr->src_ipaddr, sizeof(src_ip));
```
Do we really need intermediate variable here, can't we just do
```
net_ipv4_is_addr_loopback(UNLIGNED_GET(...));
```
here?

rlubos replied:
I don't think we could use `UNALIGNED_GET()` in this case, the function expects a pointer, but this macro returns a value, and it's not possible to obtain the address of the result, it ends up in a compilation error:
```
error: lvalue required as unary '&' operand
```
I tried however a different approach - instead of using the local copy, simply feed the function with the raw buffer address, typecasted into `in_addr *` in this case - the compiler had no issue with that. All these little address check functions use `UNALIGNED_GET()` macro internally to access the structure fields anyway, so I think it should be safe to provide it even with an unaligned address, but maybe I miss something, I don't know.

But yeah, now that all of the "hotspots" are identified, we can think of possible optimizations.

carlescufi replied:
I would say that we try to not do both, which is clearly overkill:
- if we use a buffer variable, let's not use `GET_UNALIGNED` on it
- if we use the raw packed variable, let's use `GET_UNALIGNED` on it

jukkar commented on subsys/net/l2/ethernet/arp.c:
```diff
@@ -568,6 +573,9 @@ enum net_verdict net_arp_input(struct net_pkt *pkt,
 		return NET_DROP;
 	}
 
+	net_ipv4_addr_copy_raw((uint8_t *)&src_ip, arp_hdr->src_ipaddr);
```
same question here

jukkar reviewed (commented):
This looks mostly reasonable, I am a bit worried about the extra memcpy to intermediate variable we do in various places and was wondering if we could get away with those copies by using `UNALIGNED_GET()` macro.

carlescufi reviewed (commented):
First take after a quick look

carlescufi commented on include/net/net_ip.h:
```diff
@@ -737,6 +737,18 @@ static inline bool net_ipv4_is_ll_addr(const struct in_addr *addr)
+static inline void net_ipv4_addr_copy_raw(uint8_t *dest,
+					  const uint8_t *src)
+{
+	net_ipaddr_copy((struct in_addr *)dest, (const struct in_addr *)src);
```
Scratch that, `UNALIGNED_PUT` ends up referring to the type of `dest`.

rlubos commented:
I've pushed an updated version of the PR, where I removed the `struct in_addr`/`struct in6_addr` where I found it appropriate, as requested. This also reduced the memory footprint - the `rom_report` states that after the changes the actual footprint of the `net` subsystem is actually slightly smaller, by 94 bytes.

I've rebased the PR, and resolved any new issues that showed up after the rebase. I think it's the time to pull the PR from the draft stage.

rlubos commented:
After discussing with @carlescufi we've decided to drop the warning override removal from this PR, and address remaining issues separately in other PR. This one will only focus on the networking part.
